controllers/vmware/virtualmachinegroup_reconciler.go

```diff
@@ -0,0 +1,400 @@
+// belonging to that same MD are skipped.
+func GenerateVirtualMachineGroupAnnotations(ctx context.Context, kubeClient client.Client, vmg *vmoprv1.VirtualMachineGroup, machineDeployments []string) (map[string]string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(4).Info(fmt.Sprintf("Generating annotations for VirtualMachineGroup %s/%s", vmg.Name, vmg.Namespace))
+
+	annotations := vmg.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Iterate through the VMG's members in Status.
+	for _, member := range vmg.Status.Members {
+		ns := vmg.Namespace
+		// Only VirtualMachines contribute to placement decisions.
+		if member.Kind != "VirtualMachine" {
+			log.Info(fmt.Sprintf("Member %s of %s/%s is not VirtualMachine type, skipping it", member.Name, vmg.Name, vmg.Namespace))
+			continue
+		}
+
+		// Skip it if member's VirtualMachineGroupMemberConditionPlacementReady is still not true.
+		if !conditions.IsTrue(&member, vmoprv1.VirtualMachineGroupMemberConditionPlacementReady) {
+			log.Info(fmt.Sprintf("Member %s of %s/%s is not PlacementReady, skipping it", member.Name, vmg.Name, vmg.Namespace))
+			continue
+		}
+
+		// Get VSphereMachine which share the same Name of the member Name and get the MachineDeployment Name it belonged to.
+		vsmKey := types.NamespacedName{
+			Name:      member.Name,
+			Namespace: vmg.Namespace,
+		}
+		vsm := &vmwarev1.VSphereMachine{}
+		if err := kubeClient.Get(ctx, vsmKey, vsm); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Info(fmt.Sprintf("VSphereMachine %s/%s by member Name %s is not found, skipping it", member.Name, ns, member.Name))
+				continue
+			}
+			log.Error(err, "failed to get VSphereMachine %s/%s", member.Name, ns)
+			return nil, err
+		}
+
+		mdNameFromLabel, found := vsm.Labels[clusterv1.MachineDeploymentNameLabel]
+		if !found {
+			log.Info(fmt.Sprintf("Failed to get MachineDeployment label from VSphereMachine %s/%s, skipping it", member.Name, ns))
+			continue
+		}
+
+		// If we already found placement for this MachineDeployment, continue and move to next member.
+		if v, found := annotations[fmt.Sprintf("%s/%s", ZoneAnnotationPrefix, mdNameFromLabel)]; found {
+			log.V(4).Info(fmt.Sprintf("Skipping MachineDeployment %s/%s, placement annotation %s already found", mdNameFromLabel, vsm.Namespace, v))
+			continue
+		}
+
+		// Check if this VM belongs to any of our target MachineDeployments.
+		// Annotation format is "zone.cluster.x-k8s.io/{machine-deployment-name}".
+		for _, md := range machineDeployments {
+			if mdNameFromLabel != md {
+				continue
+			}
+
+			// Get the VM placement information by member status.
+			// VMs that have undergone placement do not have Placement info set, skip.
+			if member.Placement == nil {
+				log.V(4).Info(fmt.Sprintf("VM %s in VMG %s/%s has no placement info. Placement is nil", member.Name, vmg.Name, ns))
```

Comment:
```suggestion
				log.V(5).Info(fmt.Sprintf("VM %s in VMG has no placement info. Placement is nil", member.Name))
```

the VMG will already be a k/v pair on the logger after the change suggested above

Please also use namespace/name consistently  and not sometimes use name/namespace

Similar for the next log below and the one lin l.375